astropy/io/ascii/latex.py

```diff
@@ -179,7 +187,7 @@ def write(self, lines):
 class LatexData(core.BaseData):
     """Class to read the data in LaTeX tables."""
 
-    data_start = None
+    data_start: ClassVar[str | None] = None
```

Comment:
aren't `data_end` and `splitter_class` `ClassVar`s too ?
I think we need to get this right from the get go, otherwise any class-level attribute that *isn't* annotated as `ClassVar` will be understood as an instance attribute by human readers (although it is fundamentally ambiguous but I don't think the type system currently supports explicitly declaring instance variables as such)

Reply:
The code here uses a very unusual design pattern where some class members are normal attributes in a parent class, e.g. `BaseHeader.start_line`
https://github.com/astropy/astropy/blob/76cbb5ad5dc4af1fefec84d9900e8ca03898326d/astropy/io/ascii/core.py#L577-L578
but they can be normal instance methods in a subclass, e.g. `LatexHeader.start_line`: https://github.com/astropy/astropy/blob/76cbb5ad5dc4af1fefec84d9900e8ca03898326d/astropy/io/ascii/latex.py#L136
I suspect this is impossible to express with Python's type system. At the very least Mypy keeps complaining about everything I've tried. This leaves us with three options

1) Do not annotate anything.
2) Annotate only methods, leave all attributes unannotated to avoid misleading human readers about what are class variables and what are not.
3) Annotate some but not all attributes so that a type checker could be used to verify the self-consistency of the annotations that are present.

I don't like 1), you are suggesting 2) and I went with 3) (the specific command I used was `mypy --follow-imports=silent astropy/io/ascii/{core,latex}.py`).

Reply:
thanks for the detailed explanation (and sorry it took me so long to get to it). I had not realized how complex and unusual the design was here. With this new perspective, I agree that 3 is better than 2 !